feat: SIP031 Clarity code and tests

hstove opened the pull request:
This PR builds on top of https://github.com/stacks-network/stacks-core/pull/6261, so I'm opening a draft, so that the diff is limited to just the Clarity code & tests.

The tests use [Clarigen](https://www.clarigen.dev) for easy type-safe unit tests. Clarigen is also used by sBTC and other projects. The `clarigen-types.ts` file is auto-generated based on Clarity contracts.

hstove committed:
feat: basic clarigen setup

hstove committed:
feat: initial tests for .sip-031

hstove committed:
feat: test for indirect calling `update-recipient`

hstove committed:
feat: tests/code for `claim`

hstove committed:
feat: implement logic for monthly vesting math

hstove committed:
feat: vest 100M STX over 2 years

matteojug commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+
+(define-data-var recipient principal tx-sender)
+
+(define-data-var deploy-block-height uint burn-block-height)
```
qq: is there a reason to prefer a data-var for `deploy-block-height`, instead of a constant?

obycode replied:
Agreed, this could be a constant.

obycode replied:
(it is now)

matteojug replied:
👍

matteojug commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+                    (- balance reserved)
+                    u0))
+            (vested-to-claim (if (> available-vested claimable) claimable available-vested))
+            (extra-to-claim  (- claimable vested-to-claim))
```
nit: `extra-to-claim` is not used, is it expected?

matteojug replied:
👍

matteojug commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+            (available-vested (- total-vested vested-claimed))
+            ;; Portion of the initial mint that is *still* locked (not yet vested)
+            (reserved (- INITIAL_MINT_AMOUNT total-vested))
+            ;; Free balance = everything the caller may withdraw right now
+            (claimable
+                (if (> balance reserved)
+                    (- balance reserved)
+                    u0))
```
qq/sanity check: we expect the totality of `INITIAL_MINT_AMOUNT` to be transferred to the smart contract before the first vesting?

hstove replied:
Yep - our internal stacks-core code automatically mints 200M to the contract during the epoch transition.

matteojug replied:
👍

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+    (let
+        (
+            (balance        (stx-get-balance (as-contract tx-sender)))
+            (burn-height    burn-block-height)
```
This looks unnecessary.

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+(define-private (calc-total-vested (burn-height uint))
+    (let
+      (
+        (diff (- burn-height (var-get deploy-block-height)))
```
We need to make sure that in the stacks-node changes, we deploy this contract at burn block 907740, the same height that we mint the STX (as opposed to deploying it beforehand, and only minting at block 907740. Maybe it would be safer to just hard-code that block height instead of using the deploy height?

hstove replied:
Updated in [7575db8](https://github.com/hstove/stacks-core/pull/8/commits/7575db8f7cc365870deac853089f8e59668c4b2c)

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+    (let
+      (
+        (diff (- burn-height (var-get deploy-block-height)))
+        (iterations (/ diff INITIAL_MINT_VESTING_ITERATION_BLOCKS))
```
Maybe add a comment here, so it's obvious to readers that this integer division rounds down or floors the result.

hstove replied:
Updated in https://github.com/hstove/stacks-core/pull/8/commits/41f7146cb28eb1ce1e4b1ce4b51c5cd84e5ed1fb

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+      (
+        (diff (- burn-height (var-get deploy-block-height)))
+        (iterations (/ diff INITIAL_MINT_VESTING_ITERATION_BLOCKS))
+        (stx-per-iteration (/ INITIAL_MINT_VESTING_AMOUNT INITIAL_MINT_VESTING_ITERATIONS))
```
Why don't we just make this a constant?

obycode replied:
👍

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+        (diff (- burn-height (var-get deploy-block-height)))
+        (iterations (/ diff INITIAL_MINT_VESTING_ITERATION_BLOCKS))
+        (stx-per-iteration (/ INITIAL_MINT_VESTING_AMOUNT INITIAL_MINT_VESTING_ITERATIONS))
+        (vesting-multiple (* stx-per-iteration iterations))
```
Could we call this `vested-multiple`?

obycode replied:
👍

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+        ;; If we have completed (or exceeded) the scheduled number of iterations,
+        ;; consider the *entire* vesting bucket unlocked.  This avoids leaving a
+        ;; tiny remainder caused by integer-division truncation.
+        (vesting-amount (if (>= iterations INITIAL_MINT_VESTING_ITERATIONS)
```
And this one `vested-amount`.

obycode replied:
👍

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
-(define-public (claim) (ok true) ) ;; Returns (response uint uint)
\ No newline at end of file
+;; Returns the amount of STX that is vested at `burn-height`
+(define-read-only (calc-vested-amount (burn-height uint))
```
Might be more clear to rename/reword this a bit:
```suggestion
;; Returns the amount of STX that is claimable from the vested balance at `burn-height`
(define-read-only (calc-claimable-amount (burn-height uint))
```

obycode replied:
👍

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+      (
+        (total-vested (calc-total-vested burn-height))
+      )
+      (ok (- total-vested (var-get vested-claimed-amount)))
```
Do we need this to be a response? Could it just be the `uint` only?

obycode replied:
👍

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+            (total-vested   (calc-total-vested burn-height))
+            (vested-claimed (var-get vested-claimed-amount))
+            ;; Vested that has not yet been claimed
+            (available-vested (- total-vested vested-claimed))
```
This is the same logic that is in `calc-vested-amount`. Let's just call that so we don't repeat the logic.

obycode commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+        (try! (as-contract (stx-transfer? claimable tx-sender (var-get recipient))))
+        (ok claimable)
+        )
+)
```
This function could be simplified:
```clarity
(define-public (claim)
  (let (
      (balance (stx-get-balance (as-contract tx-sender)))
      (total-vested (calc-total-vested burn-block-height))
      ;; Portion of the initial mint that is *still* locked (not yet vested)
      (reserved (- INITIAL_MINT_AMOUNT total-vested))
      ;; Free balance = everything the caller may withdraw right now
      (claimable (if (> balance reserved)
        (- balance reserved)
        u0
      ))
    )
    (try! (validate-caller))
    (asserts! (> claimable u0) (err ERR_NOTHING_TO_CLAIM))
    (var-set vested-claimed-amount total-vested)

    (try! (as-contract (stx-transfer? claimable tx-sender (var-get recipient))))
    (ok claimable)
  )
)
```

obycode reviewed (commented):
It looks pretty good, but I think some things can be simplified.

obycode commented on contrib/core-contract-tests/package.json:
```diff
@@ -11,7 +11,10 @@
+    "@clarigen/cli": "2.1.3",
+    "@clarigen/core": "2.1.3",
+    "@clarigen/test": "2.1.3",
+    "@hirosystems/clarinet-sdk": "2.16.0",
```
Does clarigen not work with the latest version of the clarinet-sdk?

hstove replied:
Not yet - it'll need a new major release to work with Stacks.JS v7. That shouldn't be hard, though.

hstove committed:
fix: code review improvements

hstove commented:
@obycode mind giving this another look?

matteojug commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,104 @@
+        )
+        (try! (validate-caller))
+        (asserts! (> claimable u0) (err ERR_NOTHING_TO_CLAIM))
+        (var-set vested-claimed-amount (+ vested-claimed total-vested))
```
Setting `vested-claimed-amount` to `(+ vested-claimed total-vested)` isn't double counting `vested-claimed`?

hstove replied:
Ah, good catch. My latest version wasn't using this var in calculations, which is why the tests were passing. I've pushed a new update that removes this var completely, and now `calc-claimable-amount` uses the same logic as what's used in `claim`.

matteojug replied:
Nice! It seems the logic is exactly the same (modulo when `balance` is fetched) -- couldn't we just call `(calc-claimable-amount burn-block-height)` in `claim` to avoid duplication?

obycode replied:
Right, this should just be `(var-set vested-claimed-amount total-vested)`.

obycode replied:
This has been removed now.

hstove replied:
Resolved in https://github.com/hstove/stacks-core/pull/8/commits/8e5292a076b8925df9a047635e6861ffe177b627

hstove committed:
fix: remove vested-claimed-amount

hstove committed:
fix: create Clarinet v1 manifest

One of our CI jobs uses Clarinet v1, which ran tests "internally", to run a single contract test. This version of Clarinet is quite old, and it doesn't support epochs newer than 2.4.

To fix this, I've created a manifest `Clarinet-v1.toml`, which is used for just running this test.

hstove committed:
fix: use separate deployment plan for v1 tests

hstove committed:
Merge branch 'feat/sip_031_boot_contract' into feat/sip31-boot-contract-tests

kantai commented on stackslib/src/chainstate/stacks/boot/sip-031.clar:
```diff
@@ -1,11 +1,106 @@
+)
+
+(define-private (validate-caller)
+    (ok (asserts! (is-eq contract-caller (var-get recipient)) (err ERR_NOT_ALLOWED)))
```
Should we assert both that contract-caller and tx-sender is the recipient? I think it only matters if the recipient is a contract.

hstove replied:
Hm.. if the recipient is a contract, it would use `(as-contract (contract-call? ...))`, in which case the `contract-caller` case is still relevant, right?

kantai replied:
Yes, but we'd also want to be sure to check the `tx-sender`, otherwise, the contract could "accidentally" invoke `validate-caller` without doing `as-contract` first.

kantai replied:
So something like this:

```suggestion
    (begin
          (asserts! (is-eq contract-caller (var-get recipient)) (err ERR_NOT_ALLOWED))
          (asserts! (is-eq tx-sender (var-get recipient)) (err ERR_NOT_ALLOWED))
          (ok true)))
```

hstove replied:
Yep, makes sense! Added in https://github.com/hstove/stacks-core/pull/8/commits/25c2b14281344577595f59bf8f1b71be45715cf4

kantai reviewed (commented):
The contract looks good to me, just had one comment on the authorization check.

hstove committed:
fix: remove duplicate code - reuse `calc-claimable-amount`

hstove committed:
Merge branch 'feat/sip_031_boot_contract' into feat/sip31-boot-contract-tests

hstove committed:
fix: underflow error in `calc-claimable-amount`

hstove committed:
fix: uppercase constant `DEPLOY_BLOCK_HEIGHT`

hstove committed:
fix: print events for important functions

hstove committed:
fix: also validate tx-sender in auth check

hstove committed:
fix: deployment path for contract tests

rdeioris committed:
minor fixes

rdeioris committed:
added check for liquidity

rdeioris committed:
Merge pull request #6261 from rdeioris/feat/sip_031_boot_contract

Feat/sip 031 boot contract

hstove committed:
Merge remote-tracking branch 'core/release/3.2.0.0.0-rc1' into feat/sip31-boot-contract-tests

hstove committed:
fix: remove `depends_on` in Clarinet.toml

moodmosaic commented on contrib/core-contract-tests/tests/sip-031/sip-031.test.ts:
```diff
@@ -0,0 +1,320 @@
+import { project, accounts } from '../clarigen-types'; // where your [types.output] was specified
```
@hstove do you run those tests in `contrib/core-contract-tests` with `npx vitest tests/sip-031/sip-031.test.ts`?

hstove replied:
yep!

hstove committed:
chore: use epoch 3.1 in simnet deployment

hstove committed:
fix: block heights in name_register.test.ts

hstove committed:
fix: test formatting

hstove committed:
fix: better resolution when calculating stx vested

hstove committed:
crc: improved cost for `update-recipient` and `validate-caller`

hstove committed:
fix: improved comments (QA-04)

hstove committed:
feat: format with `clarinet format`

hstove committed:
fix: remove redundant `begin`, QA-05

hstove committed:
fix: improve comment for `calc-claimable-amount`

hstove committed:
fix: use `is-standard` to check new recipient addr
